EIPS/eip-evm64-eof.md

```diff
@@ -0,0 +1,65 @@
+---
+title: EVM64 - EOF support
+description: EVM Object Format support for EVM64
+author: Wei Tang (@sorpaas)
+discussions-to: https://ethereum-magicians.org/t/eip-7937-64-bit-mode-evm-opcodes-evm64/23794
+status: Draft
+type: Standards Track
+category: Core
+created: 2025-05-26
+requires: 3500, 3670, 4200, 7937
+---
+
+## Abstract
+
+This EIP defines EOF support for EVM64 with its additional code validation rules and `RJUMPI`, `RJUMPV` opcodes.
+
+## Motivation
+
+EOF defines a stricter stricter code validation rule to improve efficiency. Due to EVM64 using multibyte opcode (the mode opcode `C0`), a small adaptation is needed. This EIP also additionally defines a 64-bit mode `RJUMPI` and `RJUMPV` to be 64-bit.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+We define the following gas cost constant:
+
+* `G_RJUMPIV64`: 3
+
+If EOFv1 mode is entered, then an additional validation step is added. If the opcode `C0` is encountered and it is not part of PUSH opcode's data, then the interpreter MUST validate that:
+
+* The next opcode exists.
+* The next opcode is one of the valid 64-bit mode opcode described above.
+
+For flow operations RJUMPI and RJUMPV, the 64-bit mode has following changes:
+
+* For `RJUMPI64` (0xc0e1), the condition popped from stack is only read for the last 64 bits. Gas cost is `G_RJUMPIV64`.
```

Comment:
RJUMPI checks if the condition is zero and then does not jump. If it is nonzero, it will jump to the encoded relative offset. The same goes for RJUMPV except that RJUMPV has one byte to encode how many entries there are in this table (so max 256 entries). The `case` variable from the stack is popped and if this is larger than the amount of cases encoded, it will then not jump, otherwise it will read the relative jump from the table. See also https://github.com/ethereum/EIPs/blob/master/EIPS/eip-4200.md#specification

Reply:
Also note from spec:

> The immediate argument `relative_offset` is encoded as a 16-bit signed (two's-complement) big-endian value. 

So the relative jump offset also fits in the 64bit mode. So also all (up to 256) RJUMPV table entries will fit in the 64bit mode.

Reply:
Nvm, apply the 64bit mask on the possible 256 number and then read the condition. This is indeed different.